### PR TITLE
Map navigation

### PR DESCRIPTION
**Pull request by patwolfe**

### CODE REVIEW

File: navigation.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            nodes.remove(min_node)
+            current_weight = visited[min_node]
+
+            for to_node in self.graph.edges[min_node]:
+                weight = current_weight + self.graph.getWeight(min_node, to_node)
+                if to_node not in visited or weight < visited[to_node]:
+                    visited[to_node] = weight
+                    path[to_node] = path[min_node] + [to_node]
+
+        return path[dest][1], visited[path[dest][1]], visited[dest]
```
Comment (patwolfe) [inline navigation.py]: This looks like it returns an element of the path from `curr` to `dest` (so a single node in the graph), which is then stored in the variable `path` which itself is subscripted and has its length calculated (implying it's also a list). Is that what's supposed to be happening?

File: navigation.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        self.__turnstile.release()
+
+        with self.__roomEmpty:
+            path, wait, total_wait = self.dijkstra(body["curr"], body["dest"])
+        
+        print("PATH")
+        print(path)
+    
+        if len(path) == 1:
+            road = ""
```
Comment (patwolfe) [inline navigation.py:30]: If the path is length 1, do we still need to traverse the edge (street) between the car's current node and the destination? Like if we're at node A and Packard Ave is between A and destination node B, I think you'd still want to tell the car to drive down Packard Ave.

File: navigation.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import threading
+from map import Graph
+
+class Navigation:
+    """docstring for ."""
+
+    def __init__(self, filename):
+        self.graph = Graph(filename)
+        self.__roomEmpty = threading.Lock()
+        self.__turnstile = threading.Lock()
```
Comment (patwolfe) [inline navigation.py:15]: ~~It looks like every time the map is accessed for directions a car moves from its current street to the next one in its path (so a write). Are there any reader threads in this solution? If not, we can just protect the map state with a single mutex.~~

Nevermind, I see how this works now.

File: navigation.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        with self.__roomEmpty:
+            path, wait, total_wait = self.dijkstra(body["curr"], body["dest"])
+        
+        print("PATH")
+        print(path)
+    
+        if len(path) == 1:
+            road = ""
+        else:
+            road = graph.edges[path[0]][path[1]].road_name
```
Comment (wolfep15) [inline navigation.py]: Is `Graph` subscriptable? I'm getting an error that says it's not.

**Comment by patwolfe** [comment]
I think it should be okay now?  (It runs with one car and it seems like the output is right)

**Pull request merged**